File: backend/services/notificationService.js
/**
 * Notification Service
 *
 * Centralised fire-and-forget helper used by every AI agent.
 * It saves a Notification record and optionally sends an email.
 *
 * Usage:
 *   notify(userId, { type, title, message, emailFn })
 *
 * - Always resolves immediately (never throws).
 * - Email failure is logged but never propagates.
 * - DB failure is logged but never propagates.
 */

const Notification = require('../models/Notification');

/**
 * @param {string|ObjectId} userId
 * @param {object} options
 * @param {string} options.type         - notificationType (e.g. 'coding', 'technical_interview')
 * @param {string} options.title        - short notification title
 * @param {string} [options.message]    - longer description
 * @param {Function} [options.emailFn]  - async () => Promise — call the relevant emailService fn here
 */
const notify = (userId, { type, title, message = '', emailFn = null }) => {
  // Run everything asynchronously — caller never awaits this
  (async () => {
    let emailSent = false;

    // 1. Send email first (if provided)
    if (typeof emailFn === 'function') {
      try {
        await emailFn();
        emailSent = true;
      } catch (err) {
        console.error(`[NotificationService] Email failed (type=${type}):`, err.message);
      }
    }

    // 2. Save notification record
    try {
      await Notification.create({
        userId,
        notificationType: type,
        title,
        message,
        status: 'unread',
        emailSent
      });
    } catch (err) {
      console.error(`[NotificationService] DB save failed (type=${type}):`, err.message);
    }
  })();
  // Return nothing — fully fire-and-forget
};

module.exports = { notify };

File: backend/services/notification_service.py
"""
services/notification_service.py — Notification helper (replaces notificationService.js)
Fire-and-forget: saves notification to DB + optionally sends email.
"""
import asyncio
import logging
from sqlalchemy.orm import Session
from models.notification import Notification

logger = logging.getLogger(__name__)


def notify(
    db: Session,
    user_id: int,
    notification_type: str,
    title: str,
    message: str = "",
    email_fn=None,
):
    """
    Fire-and-forget notification helper.
    Saves a Notification row and optionally calls an async email function.
    Never raises — all errors are logged only.
    """
    async def _run():
        email_sent = False
        if email_fn is not None:
            try:
                await email_fn()
                email_sent = True
            except Exception as e:
                logger.error(f"[NotificationService] Email failed (type={notification_type}): {e}")

        try:
            notif = Notification(
                user_id=user_id,
                notification_type=notification_type,
                title=title,
                message=message,
                status="unread",
                email_sent=email_sent,
            )
            db.add(notif)
            db.commit()
        except Exception as e:
            logger.error(f"[NotificationService] DB save failed (type={notification_type}): {e}")
            db.rollback()

    # Schedule as a background task using asyncio
    try:
        try:
            loop = asyncio.get_running_loop()
            loop.create_task(_run())
        except RuntimeError:
            asyncio.run(_run())
    except Exception as e:
        logger.error(f"[NotificationService] Failed to schedule: {e}")
